src/coreclr/vm/loaderallocator.cpp

```diff
@@ -2198,3 +2205,250 @@ PTR_OnStackReplacementManager LoaderAllocator::GetOnStackReplacementManager()
+{
+    STANDARD_VM_CONTRACT;
+    g_pMoveableGCPointerTracker = new CollectibleMoveableGCPointerTracker();
+}
+
+CollectibleMoveableGCPointerTracker::CollectibleMoveableGCPointerTracker() : 
+    m_pointers(nullptr),
+    m_numPointers(0),
+    m_maxPointers(0),
+    m_nextTableScanChunk(-1),
+    m_Crst(CrstLeafLock, (CrstFlags)(CRST_UNSAFE_COOPGC | CRST_REENTRANCY))
+{
+}
+
+void CollectibleMoveableGCPointerTracker::AddPointer(OBJECTHANDLE gcHandleDescribingLifetime, Object *pObject, uintptr_t *pTrackedInteriorPointer)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE; // Since the GC reads the array without taking locks, we need to be in cooperative mode
+        INJECT_FAULT(COMPlusThrowOM());
+    }
+    CONTRACTL_END;
+
+    CrstHolder ch(&m_Crst);
+    MoveableGCPointer structToAdd;
+    structToAdd.LifetimeTrackingHandle = gcHandleDescribingLifetime;
+    structToAdd.PointerValue = pObject;
+    structToAdd.TrackedInteriorPointer = pTrackedInteriorPointer;
+    if (m_numPointers >= m_maxPointers)
+    {
+        uintptr_t newMaxPointers = max(16, m_maxPointers * 2);
+        MoveableGCPointer *newPointers = new MoveableGCPointer[newMaxPointers];
+        if (m_pointers != nullptr)
+        {
+            memcpy(newPointers, m_pointers, m_maxPointers * sizeof(MoveableGCPointer));
+            delete[] m_pointers;
+        }
+        m_pointers = newPointers;
+        m_maxPointers = newMaxPointers;
+    }
+    
+    m_pointers[m_numPointers] = structToAdd;
+    m_numPointers++;
+}
+
+void CollectibleMoveableGCPointerTracker::ScanTable(promote_func* fn, ScanContext* sc)
+{
+    // This executes during the GC, so we don't need to take a lock, as the table cannot be modified, as all modifications involve cooperative mode
+    WRAPPER_NO_CONTRACT;
+
+
+    if (sc->promotion)
+    {
+        // We must not promote anything, as the lifetime of these pointers is driven by the lifetime of the LifetimeTrackingHandle
+        return;
+    }
+
+    if (m_numPointers == 0)
+    {
+        // There isn't any work to do.
+        return;
+    }
+
+    // On the server heap, where multiple threads may be scanning the table, we can to divide the work up. Use a simple work-stealing approach
+    //
+    // Notably, in GCToEEInterface::AnalyzeSurvivorsFinished which is called once for each time that there is a set of relocations, we reset the current
+    // chunk index to -1. Then in this function, we operate on chunks from 0 to maxChunkCount. We use InterlockedIncrement to get the next chunk index, and
+    // when a chunk is done, we increment the next chunk index. This is done is a loop, so if 1 (or more) threads of the GC is running ahead of the others,
+    // it will do more chunks than another thread which is running behind, thus balancing the work.
+    uintptr_t maxChunkCount = GCHeapUtilities::IsServerHeap() ? sc->thread_count : 1;
+    LONG chunkIndex = InterlockedIncrement(&m_nextTableScanChunk);
```

Comment:
Possibly. Clearly it could be, and it will certainly show up in VTune, but implementing things this way provides a scheme for GC threads that are running ahead of other threads to do all the work before the rest of the threads finish up. My understanding is that GC threads are rarely perfectly balanced, but I'd like to hear the opinion from @Maoni0 here.

Reply:
Notably, there is straightforward alternative approach where each chunk is directly mapped to one of the GC threads. Then we don't need a `m_nextTableScanChunk` variable and there is no thread contention. Given that this is only a cost in the presence of collectible statics I could see us doing either of these.

Reply:
it's very difficult to get the GC threads perfectly balanced but we always try to get them as balanced as possible and in the past we already ran into perf issues where the EE side of scanning is very costly due to taking locks so this is certainly we should measure and make sure we do a reasonably good job at.